samcli/local/lambdafn/runtime.py

```diff
@@ -161,28 +242,278 @@ def _get_code_dir(self, code_path):
+        self._lock = threading.Lock()
+        super().__init__(container_manager, image_builder)
+
+    def create(self, function_config, debug_context=None, event=None):
+        """
+        Create a new Container for the passed function, then store it in a dictionary using the function name,
+        so it can be retrieved later and used in the other functions. Make sure to use the debug_context only
+        if the function_config.name equals debug_context.debug-function or the warm_containers option is disabled
+
+        Parameters
+        ----------
+        function_config FunctionConfig
+            Configuration of the function to create a new Container for it.
+        debug_context DebugContext
+            Debugging context for the function (includes port, args, and path)
+        event str
+            input event passed to Lambda function
+
+        Returns
+        -------
+        Container
+            the created container
+        """
+
+        # reuse the cached container if it is created
+        container = self._containers.get(function_config.name, None)
+        if container and container.is_created():
+            LOG.info("Reuse the created warm container for Lambda function '%s'", function_config.name)
+            return container
+
+        # debug_context should be used only if the function name is the one defined
+        # in debug-function option
+        if debug_context and debug_context.debug_function != function_config.name:
+            LOG.debug(
+                "Disable the debugging for Lambda Function %s, as the passed debug function is %s",
+                function_config.name,
+                debug_context.debug_function,
+            )
+            debug_context = None
+
+        container = super().create(function_config, debug_context, None)
+        self._containers[function_config.name] = container
+        self._add_function_to_observer(function_config)
+        return container
+
+    def _on_invoke_done(self, container):
+        """
+        Cleanup the created resources, just before the invoke function ends.
+        In warm containers, the running containers will be closed just before the end of te command execution,
+        so no action is done here
 
-            elif os.path.isfile(code_path) and code_path.endswith(self.SUPPORTED_ARCHIVE_EXTENSIONS):
+        Parameters
+        ----------
+        container: Container
+           The current running container
+        """
+
+    def _add_function_to_observer(self, function_config):
+        """
+        observe the function code path, so the function container be reload if its code got changed.
 
-                decompressed_dir = _unzip_file(code_path)
-                yield decompressed_dir
+        Parameters
+        ----------
+        function_config: FunctionConfig
+            Configuration of the function to create a new Container for it.
+        """
+        with self._lock:
```

Comment:
How does this work for multiple functions getting updated at the same time? Why do we need a lock?